OneSignal/OneSignal-Android-SDK
# Read-Your-Write Consistency

# Description

## One Line Summary
Introduce and integrate a ConsistencyManager to manage read-your-write tokens for improved segment membership calculation.

## Details

### Motivation
This update introduces the **ConsistencyManager** to manage kafka offsets, which are used as read-your-write tokens. The goal is to improve the accuracy of segment membership calculations by providing an open-ended & highly customizable blocking mechanism for operations that rely on having successfully synchronized client & server state.

For a first use-case, we want to block the fetching of IAMs until we have offsets for a user & subscription state update.

### Scope
- Registers the **ConsistencyManager** in the User and InAppMessages modules.
- Updates the Subscription and User backend services to return offsets needed by the **ConsistencyManager**.
- Modifies Subscription and UpdateUser operation executors to set offsets in the **ConsistencyManager**.
- Updates the IAM manager and backend service to include the offset as a query parameter in the `listInAppMessages` request.

# Testing
## Unit testing
- New coverage for new `ConsistencyManager` class

## Manual testing
Tested on Android emulator
- Set up Android test app in OneSignal
- Note the current app version (e.g. 1.0)
- Create a segment targeting a specific version (e.g. 1.1)
- Create an in-app using the new segment
- Open the app (see you don't get an in-app)
- Close the app
- Change the version to the one in the segment (e.g. 1.1)
- Open the app (see you get the in-app)

# Affected code checklist

# Checklist
## Overview
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR

## Testing

## Final pass
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/OneSignal-Android-SDK/2168)
